### ISSUE TITLE
Graph Blocks appearance style

### ISSUE THREAD

**Issue by dube1503** [txst-per-group/Glowscript-Blocks#57]

**Comment by eah13** [comment]
I assume the question here is about the display block since the others have a selection.

#1 seems non-idiomatic, since 'hollow' blocks indicate loops.

So I guess we're choosing between manually selecting the attributes via the mutator or having some attributes pre-selected.  Pre-selecting attributes seems like a good move, especially if they're filled with the defaults.  That would be my vote.  Any edge case attributes could still be selected via the gear.

This may give rise to a useful design principle: for blocks with gear icons, pre-select common attributes and fill them with default blocks.  This makes common attributes more discoverable and the process to e.g. make a sphere with a certain radius is grab block > change radius int block instead of grab block > add radious attr > grab int > change int.  As for what should be pre-filled and what not, that would need some more discussion.

Sorry to scope creep this issue into something bigger.  We can open another issue to contain that discussion unless it seems like an auto accept or auto reject suggestion.

Nice use of paper prototyping, btw,

**Comment by codestar12** [comment]
@dube1503 and I have been talking about making the inputs that are added with gear mutators populate with "shadow" blocks to save the user time for the most common tasks and still allow pain free use of variables for the constructors.

**Comment by codestar12** [comment]
In addition we were going to start with finishing the plot and line blocks because those are all that are needed make a basic graph and the least costly to modify if we need to make design changes

**Comment by dube1503** [comment]
@eah13 I agree that option one references the style of the loop locks which would be confusing.
I would also have to agree with preselection for the attributes (and as a future design principle) as it helps ease new users into how the functionality of the program works and promotes easy discoverability, and as @codestar12 had mentioned we have had discussion on the best way to do this that meets the right blend of "easy usability" and "mildly frustrating to promote learning through exploring/tinkering"

Another issue we are trying to solve that may not have been expressed from the prototyping was somehow showing a "Display" blocks ownership of "Line" blocks (in glowscript any "Line" declarations that are created below the creation of a graph "Display" will be plotted on said Display be default when calling the Plot function for the Lines). This was the reasoning for the Statement input style block of Option 1 where Line blocks would be placed inside as @codestar12 @hgclose and myself had felt it was important to somehow show ownership over Line blocks, yet it ultimately seems a confusing allusion to other Loop style blocks. 
I have discovered however that it is possible to perform a setCheck on top and bottom block connectors so a possible solution may be that when a Display block is created, it is required the next block connected must be at least one Line type. Some obvious problems with this however are that without instruction this would not be apparent and could cause confusion as well the bypass to this could be that the user is not required to connect all blocks together to complete or run a program.

I am wondering if this is even a problem can be easily solved with the design of blocks without some sort of guidance or instruction, which seems to be the nature of learning a programming language in any shape or form.

**Comment by hgclose** [comment]
COMMENTS ON SEMANTICS OF LABELS

Note that BARS is not a type of LINE, so there needs to be a more general word than _line_. I think it is quite unfortunate that GS defines _graph_ as a set of axes, which we in our project are calling _display_. In our case then, a _graph_ is a general word for a "g-object" like gcurve, gbars, etc. Or maybe we should call it _graph types_ so that the word "Graphs" can be reserved to refer to the general category that includes "graph types", "plotting", (not "plot", so as to avoid the sense of "plot" as a noun and synonym of "graph"), and "display". 

COMMENTS ON "LINES"

I see no reason for the block to be more complicated than LINES # 1. I think I just don't understand what LINES # 2 is doing. 

COMMENTS ON PLOTTING

I prefer the ordered pair rather than referring to the values as "x" and "y". On the one hand, the variables might not be named x and y at all. On the other hand, GlowScript refers to xmin, ymax, etc. So there is no way to avoid x and y altogether. I don't know what the right solution is. Also in GS syntax, the graph is identified first, and then the point to be plotted is identified. I guess by my logic I should prefer a PLOT block that is a hybrid of 2 and 3, in which the graph is first identified as in 3, the word "plot" appears, and then the point is identified as an ordered pair as in 2. This follows the structure of the GS plot command. 

COMMENTS ON DISPLAYS

Regarding the idea of a hollow block indicating a loop: That might be right. But when Cody and I were discussing it, I don't think that it occurred to us that putting a block inside another block invoked the idea of "loop" necessarily. In my mind at least, the metaphor was more general: Something is inside something else and therefore belongs to it, or is governed by it. Like, North Carolina is conceived as being "inside" the USA container and is therefore subordinate to it in some general sense. It may turn out in user testing to be one way or the other. I would be curious actually to find out. So I do still like DISPLAY # 1. My prediction is that users will understand # 1 because the actual graphs are construed by students as being "inside" the space of the axes, and the hollow/filled display block would mirror this construal. I recognize that DISPLAY # 2 is more like GS code. I don't like how the subordinate relation of specific graphs to display is still hidden; as in, the user must look up the documentation on graphs in order to learn how graphs get assigned to displays. 

The default values of xmin, xmax, ymin, ymax are dynamic, i.e., they rescale to fit the graph as it grows. Therefore I think we should avoid setting defaults for these values with shadow blocks. So I would avoid DISPLAY # 3.

If we do DISPLAY # 1, what would happen if the hollow block were left empty? I don't think it would cause an error in GS to make a display and plot nothing. So, can we make it so the display block # 1 would be harmless if empty?

**Comment by codestar12** [comment]
@hgclose 

COMMENTS ON "LINES"

I think we need to choose something whether it be "line" or "graph type" quickly. Reading your message and discussing solutions is incredibly confusing in what "namespace" we are referring to "graph" . It carries way too much baggage to try and rebrand the word. 

Also I should point out we had decided not to include bar so far in the project so the name line shouldn't be a problem. Also there will be no label that says 'line' it is just what we are calling the block for programming it

COMMENTS ON DISPLAYS

I think it is going to be impossible for use to predict ahead of time which method will be easiest for users to understand. If we do prototyping we will need to do some paper prototyping because it will be very time consuming to have to reconfigure these graph blocks.

While I think #1 will be easy for students to understand the idea of the displays ownership over the g-types I think it will create some confusion about what the syntax relationship a block with that shape has. Every other use of that shape is equivalent to some function that is indented. Also I'm not sure that this clarity is worth breaking our design principle #13 (that only explicitly states text but I think the idea carries). 

I say we pick 2-3 ideas and have cole work on some paper prototyping so we can get feedback from users. I would prefer any solution that maps direct to code as close as possible (even if we don't like how glowscript code is written) and isn't terribly received by the user.

**Comment by hgclose** [comment]
Maybe it would helpful for me to try to indicate more clearly what I think should happen without any unnecessary musings. I don't want to say anything confusing. 
MENU: Label it "Graphs"
LINES: Use Block # 1, with "bars" omitted, including only "curve" and "dots".
PLOT: Use Block # 4, but with an ordered pair in parentheses "( , )" instead of the format "x:, y:" 
DISPLAY: Use Block # 2, with a mutator for editing xmin, xmax, ymin, ymax, and title as attributes. Pre-select the title attribute and pre-load a shadow text box attached to the attribute. Make the text box empty by default.

Others are welcome to express their disagreement with my recommendations.

Yes, I am changing my tune about the DISPLAY block because I do not see anyone agreeing with my thinking about the hollow block. I do not think we should try to get user testing in at this point before release, and while I am away from TX. If the reason for suggesting user testing is to decide on whether to make the block hollow, and if I am the only one who thinks a hollow block is a good idea, then I can just back off of the idea. Then the issue is resolved.

If there is a choice that @codestar12 thinks I should make quickly and which I have not made, please let me know. I did not understand @codestar12 's top paragraph clearly in the most recent comment. I believe the top part of this comment addresses the issues we need addressed.

**Comment by dube1503** [comment]
@hgclose I think all your comments and feedback were very helpful and enlightening (getting input from everyone on the project is extremely helpful for cross-discipline perspective in my opinion) and I dont think @codestar12 was asking for a concrete "solution" from you as much as just wanting to open a discussion for possible iterations on the design. for me the most helpful parts are when we can get feedback from every team member because alot of times an idea is presented that someone has not proposed or thought and having every option is what i think is what we need. 
The idea of the paper prototype testing was one of my advocations as it is a way for us to quickly and easily get real user feedback from our proposed solutions based on the teams thoughts, but it doesnt require any code. The user testing we were talking about was only simple one on one test with a "sponsored user" where we could put our current solutions to the design issue and get instant feedback, make those changes on paper, and then re-present them, that way we can come to our "final" solution and and iterate through possible outcomes as quickly as possible without having to dive into code and then rewrite the code.

### DISPLAY BLOCK
I want to continue to iterate and prototype the design of the "display" block more and see if there is a design or solution we have not explored yet. Although it does seem that the consensus is to avoid the hollow block as it mimics the and communicates a different "code mechanism" that what we want, i am not convinced that we have solved the problem yet.
This block is the main culprit that i want to do some paper prototyping and iteration for, i want to make sure that we explore every option because there may be a solution not presented or discussed here and i think user insight might lead us to that solution more easily and possible solve other issues we may not have thought of solutions for and i think it is better to find those solutions as quickly as possible now than to try and retroactively fix them later.

And i think a lot of the confusion, and @codestar12 top comment, have to do with the "SEMANTICS" of the blocks you had discussed. Reading through your original post was confusing at first as the word "graph" is used somewhat differently from each person and each scenario. I think our best solution to semantics may be in line with what you had originally posted in that thinking of all of these blocks as "graph elements" as in none of them are actually a "graph" (despite how they are named in Glowscript) and that only when all of the elements are combined together do they make a complete graph or Graph category. That being said, my wording and labels were only for my sake to separate them out while trying to avoid the word "graph" within one element or type and i see now may have led to some of the confusion. Some alternate proposals could be using the names G OBJECTS, PLOTTING (or PLOTTER), and DISPLAY

**Comment by codestar12** [comment]
@hgclose as far as the display block goes I also don't think we should abandon the graph with a hollow (statement) block just because no one is agreeing. We are suggesting user testing because it is very time consuming the change how these would work and it is much cheaper to get feedback before we start work on it. We also want to go into the process without bringing our own biases about what is least confusing to the user because we already know how blocks work.

**Comment by dube1503** [comment]
Yes, @codestar12 makes a good point that I did not @hgclose, just because we believe it might be confusing or that we think it doesnt make sense may not translate into what our users might think and while I don't think it is the best solutions if doing quick testing shows that it is the one that users respond to best then i would say that does make it the best solutions.
That being said, some other solutions to this issue might also be 
 - requiring a "display" block to have a "g_object" attached below it (and a way to enforce this could be having all blocks be disabled until they are all connected, leaving no "orphan blocks"
 - make the "display" block disabled by default until a "g_object" is attached below it
 - "display" block defaults with a "g_object" already attached below it
 - adding a mutator to the "g_object" block which allows dragging a "display" assignment parameter

any input from anyone on which, if any, of these additional objects could benefit from testing with the current options is needed

**Comment by eah13** [comment]
OK. After reading all of the discussions and thinking more on this I have a paper prototype for discussion. 

My strongest recommendation is to follow the Glowscript API in using the term 'series'. Series are all that are needed to plot something: they can use an anonymous/default Graph object if you call them directly. The Graph object is only needed to customize or produce multiple graphs. 

So, here's my hopefully constructive set of suggestions:

![img_20170628_081036](https://user-images.githubusercontent.com/1702745/27636464-b248b55e-5bd9-11e7-9290-2e4469e155f0.jpg)

Comments welcome.

**Comment by eah13** [comment]
A common theme: I've tried to include 'friendly' attributes like title and color in the default blocks. This can help buikd confidence by letting users easily change these properties and see the results of their changes. That implicitly teaches them what blocks correspond to which elements of the final graph. Pedagogically, I think this warrants giving them first class status in the UI even though Glowscript doesnt particularly care about them. 

Also I've shown both Series and Graph objects coming with a 'set' block by default, since I don't see a case where users would need them independently. 

If we go with this style, future calls to 'set' on these variables may need to offer the option to change these attributes, the equivalent of `graph.xmin = 4`.

**Comment by dube1503** [comment]
I like your idea of 'friendly' attributes, it seems to mimic some other block styles and was the style we also seemed to adopt for the shapes blocks as well. The mutator on the g_object, series blocks that contains a reference to the display is a nice idea as well as it mimics the code where you objects are assigned to a default display without creating one as well as the fact that you can also assign a display from the g_object too. 

It does appear to me, and i will have to get @hgclose to clarify for me as he has more experience, that the use of series here is not fully correct. While series is all that is referenced in the glowscript documentation (this is one reason i had noted in the slack channel that i believe the documentation to be quite less than perfect) they are indeed different from what we are calling "g_objects" which include gcurve, gdots, gbars _(a breif aside, from researching this topic before when first going through solutions for these blocks i discovered that these g_objects are leftover features from vpython and series are from glowscript [yet gcurves and such are still the preferred use](https://groups.google.com/forum/#!topic/glowscript-users/QZY-b5DCZ8Y))_ it is quite unfortunate to me that an entire in depth list of features is missing from the documentation and that it is not very transparent with even the existing shown features, in my opinion (a personal grub here is that the Glowscript Documentation doesn't seem to follow the standard of listing features/functions, a description, and an example which i believe it is in need of). The series i think though was created as a way of plotting sets and lists which was something we had been discussing which might show that having a series block along side the existing g_object block might be useful?
It might be helpful now to ask, do we need  continue on using the g_objects style or go with the 'Glowscript' style of using the series features, or try to integrate both?

Also a good point about the 'set' references. It seems regardless of direction we need to include that ability to 'get' and 'set' these attributes after assignment, exactly the way that the shape and vector blocks work. I predict also that some users might find this ability to reassign attributes an insight into the underlying structure of these components they are using.

**Comment by dube1503** [comment]
@eah13 not sure if you could advise, I made a slack channel for more in depth discussion as to avoid filling up the github page too much as well as being able to take advantage of slacks notifications, but I am not sure if that is the best option to discuss feedback and solutions?
